pipes-and-filters/Pipes-And-Filters/Enricher/src/main/java/Enricher.java

```diff
@@ -5,54 +5,34 @@
 import java.util.function.Function;
 
 public class Enricher {
-    public static void main(String[] args) {
+    public static void main(String[] args) throws IOException {
         IAmAnOperation<Greeting, EnrichedGreeting> greetingEnricher = new GreetingEnricher();
 
         Function<String, Greeting> greetingDeserializer = messageBody -> {
             try {
                 return new ObjectMapper().readValue(messageBody, Greeting.class);
             } catch (IOException e) {
-                e.printStackTrace();
-                return null;
+                throw new RuntimeException(e);
             }
         };
         Function<EnrichedGreeting, String> enrichedGreetingSerializer = enrichedGreeting -> {
             try {
                 return new ObjectMapper().writeValueAsString(enrichedGreeting);
             } catch (IOException e) {
-                e.printStackTrace();
-                return null;
+                throw new RuntimeException(e);
             }
         };
 
-        var consumer = new Filter<>(greetingEnricher, greetingDeserializer, enrichedGreetingSerializer, "greeting", "enrichedgreeting", "localhost");
+        var enricher = new Filter<>(greetingEnricher, greetingDeserializer, enrichedGreetingSerializer, "greeting", "enrichedgreeting", "localhost");
 
-        var executorService = Executors.newSingleThreadExecutor();
-
-        try {
+        try (var executorService = Executors.newSingleThreadExecutor()) {
             System.out.println("Enricher running, entering loop until signaled");
             System.out.println(" Press [enter] to exit.");
             // has its own thread and will continue until signaled
-            var task = consumer.run(executorService);
-
-            while (true) {
-                // loop until we get a keyboard interrupt
-                if (System.in.available() > 0) {
-                    // Note: This will deadlock with System.out.println on the task thread unless we have called println first
-                    char key = (char) System.in.read();
-                    if (key == '\n') {
-                        // signal exit
-                        task.cancel(true);
-                        break;
-                    }
-
-                    Thread.yield();
-                }
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
-        } finally {
-            executorService.shutdown();
+            var future = executorService.submit(enricher);
```

Comment:
Yeah, I did wonder if it made more sense to have the executor submit the enricher in main, than have run submit the consumer. Let me ponder how that changes what I am trying to explain. But it is a valid point